File: androidApp/src/main/java/com/demo/bookclubkmp/android/ui/screens/auth/AuthViewModel.kt
package com.demo.bookclubkmp.android.ui.screens.auth

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.demo.bookclubkmp.domain.entities.Session
import com.demo.bookclubkmp.domain.usecases.auth.ISignInUC
import com.demo.bookclubkmp.ui.viewmodels.IAuthViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class AuthViewModel (private val signInUC: ISignInUC) : ViewModel(), IAuthViewModel {
    data class UIState(
        val session: Session? = null,
        val exception: Any? = null,
        val isLoading: Boolean = false
    )

    private val _uiState = mutableStateOf(UIState())
    val uiState = _uiState

    override
    fun signIn(username: String, password: String) {
        _uiState.value = _uiState.value.copy(isLoading = true, exception = null)
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val session = signInUC.execute(username, password)
                _uiState.value = _uiState.value.copy(session = session, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(exception = e, isLoading = false)
            }
        }
    }
}

File: iosApp/iosApp/ui/screens/auth/AuthViewModel.swift

import shared

@Observable
class AuthViewModel : IAuthViewModel {
    
    init(signInUC: ISignInUC) {
        self.signInUC = signInUC
        self.session = nil
        self.exception = nil
        self.isLoading = false
    }
    
    var session: Session?
    var exception: Any?
    var isLoading: Bool
    
    let signInUC: ISignInUC
        
    func signIn(username: String, password: String) {
        isLoading = true
        signInUC.execute(username: username, password: password) { session, error in
            self.isLoading = false
            if (error != nil) {
                self.session = nil
                self.exception = error
            } else {
                self.exception = nil
                self.session = session
            }
        }
    }
}
